torchvision/prototype/transforms/functional/_color.py

```diff
@@ -227,7 +194,57 @@ def equalize_image_tensor(image: torch.Tensor) -> torch.Tensor:
     if image.numel() == 0:
         return image
 
-    return _equalize_image_tensor_vec(image.reshape(-1, height, width)).reshape(image.shape)
+    # The algorithm for histogram equalization is mirrored from PIL:
+    # https://github.com/python-pillow/Pillow/blob/eb59cb61d5239ee69cbbf12709a0c6fd7314e6d7/src/PIL/ImageOps.py#L368-L385
+    # Although PyTorch has builtin functionality for histograms, we are not using them here since they can only be
+    # applied to the complete tensor. We need to apply them to the spatial size, i.e. the last two dimensions only, and
+    # thus we would need to call them in a for loop, which is inefficient.
+    batch_shape = image.shape[:-2]
+    flat_image = image.flatten(start_dim=-2).to(torch.long)
+
+    # The histogram is computed by using the flattened image as index. For example, a pixel value of 127 in the image
+    # corresponds to adding 1 to index 127 in the histogram.
+    hist = flat_image.new_zeros(batch_shape + (256,), dtype=torch.int32)
+    hist.scatter_add_(dim=-1, index=flat_image, src=hist.new_ones(1).expand_as(flat_image))
+    cum_hist = hist.cumsum(dim=-1)
+
+    # The simplest form of lookup-table (LUT) that also achieves histogram equalization is
+    # `lut = cum_hist / flat_image.shape[-1] * 255`
+    # However, PIL uses a more elaborate scheme:
+    # 1. Instead of normalizing the cumulative histogram by the number of pixels (` / flat_image.shape[-1]` above),
+    #    it is normalized by the number of pixels that are don't have the maximum value. Note that maximum value here
+    #    does not mean 255 per se but rather the maximum value in the image, which might be or not be 255.
+    # 2. Instead of normalizing just the cumulative histogram, a constant based on the number of non-maximum is added
+    #    to it.
+    # This brings the computation of the  LUT to
+    # `lut = ((cum_hist + num_non_max_pixels // (2 * 255)) // num_non_max_pixels) * 255`
+
+    # The last non-zero element in the histogram is the first element in the cumulative histogram with the maximum value
+    index = cum_hist.argmax(dim=-1)
+    num_non_max_pixels = flat_image.shape[-1] - hist.gather(dim=-1, index=index.unsqueeze_(-1))
+
+    # This is performance optimization that saves us one multiplication later. With this, the LUT computation simplifies
+    # to `lut = (cum_hist + step // 2) // step` and thus saving the final multiplication by 255 while keeping the
+    # division count the same. PIL uses the variable name `step` for this, so we keep that for easier comparison.
+    step = num_non_max_pixels.div_(255, rounding_mode="floor")
+
+    # Although it looks like we could return early if we find `step == 0` like PIL does, that is unfortunately not as
+    # easy due to our support for batched images. We can only return early if `(step == 0).all()` holds. If it doesn't,
```

Comment:
nb. We don't do this because looking into a tensor on CUDA incurs on a perf hit due to having to synchronise the devices. If we just care about CPU and this is a very common case (you guys should decide this) then perhaps this optimisation makes sense.

Reply:
> If [...] this is a very common case

No, and it seems I didn't make it clear enough in the comment. `step == 0` can only happen in two cases:

1. The image only has a single value, meaning there are no pixels that don't have the maximum value. No real image will ever hit this  since the image contains no information whatsoever.
2. The image has multiple values, but the number of non-maximum valued pixels is lower than 255. I think this is an unintended side effect of the PIL algorithm. In a real use case, I think we should also never hit such an image. It either has to be really small like 16 x 16 or we have another unrealistic distribution of values where most of the pixels have one high value.

Reply:
To be fair, since `step.numel()` is pretty low (only the batch size), I think this check should indeed be pretty fast on CPU. We could add something like

```py
no_equalization = ...
if image.device == "cpu" and no_equalization.all():
    return image
```

But that wouldn't simplify the code that comes after that.

Reply:
you need to iterate through all the data to do the `all()` reduction, so, given that it's not a common case, I would suggest against it